### ISSUE TITLE
ngrok - consoleui failed: Failed to initialize terminal: invalid argument

### ISSUE THREAD

**Issue by mariotoffia** [Microsoft/WSL#1951]
Hi I'm trying to get ngrok working due to nodejs development and sharepoint online and webhooks. But it seems only to work on the CMD and not on the WSL. Is there any obvious that I need to do in order to run ngrok. It works very well in the commandline.

Cheers,
 Mario

* Your Windows build number:  
Microsoft Windows [Version 10.0.15063]
Ubuntu version 16.04.1 LTS

* What you're doing and what's happening: 
ngrok http 3000
(installed via `npm i -g ngrok`)

* What's wrong / what should be happening instead:
consoleui failed: Failed to initialize terminal: invalid argument

Expecting ngrok tunnel to be set up like it would in the CMD.

* Strace of the failing command, if applicable: 
Multiple strace.txt so I've zipped them as a attachment.
[strace.zip](https://github.com/Microsoft/BashOnWindows/files/933818/strace.zip)

**Comment by devchakraborty** [comment]
Same issue - Windows 10.0.15063, Ubuntu 16.04.

**Comment by mabdelfattah** [comment]
+1

**Comment by Apoorv97** [comment]
+1 
same here

**Comment by kogeva** [comment]
+1

**Comment by aaronleslie** [comment]
same to me, we are under company network

**Comment by sqlsolver** [comment]
+1

**Comment by tomjw64** [comment]
I had the same issue on WSL. Using ngrok's `--log` option to write the log to stdout or stderr solved the problem for me. e.g. `ngrok http 3000 --log stderr`

**Comment by HarrySarson** [comment]
The `--log`  work around works for me :)

**Comment by sunilmut** [comment]
@all - Apologize for the delay in getting to this and thanks for the report & traces. I looked at the traces. The problem seems to be coming from the lack of support for signal driven I/O in WSL. Here are some snippets from the traces that establishes that. Also adding @SvenGroot as FYI.

```
$ grep 'EINVAL' *
strace.txt.76:fcntl(4, F_SETOWN, 74)                  = -1 EINVAL (Invalid argument)

$ grep '= 4' *
strace.txt.74:socket(PF_INET6, SOCK_STREAM, IPPROTO_TCP) = 4
strace.txt.74:openat(AT_FDCWD, "/dev/tty", O_RDONLY)  = 4
```

And, in the case `ngrok http 3000 --log stderr` where it doesn't give out the error, there is not such `fcntl` issued. I would recommend opening up a [user voice ticket](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/category/161892-bash) for signal driven I/O as it is a feature level request.

Regarding `tunnel does not start`, the web service starts fine for me at `127.0.0.1:4040`. How can I validated whether the tunnel started or not?

**Comment by HarrySarson** [comment]
Sorry, I just panicked when the command didn't give me a url to go to.

Have tried again and the web service works fine, many apologies.

**Comment by sunilmut** [comment]
@HarrySarson - No problems.

Just FYI that here is another request for signal I/O #601

**Comment by mgcox2** [comment]
I used the following options and I was able to see the public URL in the debug output, and it did in fact work. I was able to browse to it on my Android phone:
`--log "stdout" --log-level "debug"`

**Comment by abdilahrf** [comment]
same issue 👍

**Comment by iamfaith** [comment]
same issue.
hope solve soon.

**Comment by therealkenc** [comment]
Ref #601

**Comment by rfgamaral** [comment]
What's the status on this and does anyone now if anything is being done to fix this issue?

**Comment by Brian-Perkins** [comment]
There have been changes in Insiders Build to address issues like signal-driven IO support for tty/pty (see therealkenc's ref above) so it would be interesting to try this on a recent Insiders release. If there are still issues then we need to dig into the strace, et al. to see what is still blocking.

**Comment by rfgamaral** [comment]
My machine has been very unstable as of late otherwise I'd by willing to install the Insiders build and give it ago. Hopefully someone else can do it and report back!?

**Comment by rfgamaral** [comment]
Looking at the issue @therealkenc referenced and the [Release Notes on Build 17017 ](https://docs.microsoft.com/en-us/windows/wsl/release-notes#build-17017) should have this fixed?  I mean, there's this:

> O_ASYNC support for tty/pty files.

Then again, I have no idea what I'm talking about and I'm just trying to put the pieces together.

tl;dr;
Is it possible that the Windows Insider Build 17017 fixes this issue?

**Comment by rfgamaral** [comment]
Decide to install the latest downloadable Insider Preview ISO and test this for myself and it's working :)

![image](https://user-images.githubusercontent.com/96476/35114105-8eb39c7e-fc7b-11e7-9d6c-42c7b136a1b6.png)

Now I'm thinking on upgrading to Insider Preview on my main system, but afraid that it might not be very stable...

**Comment by benhillis** [comment]
Marking as fixed in 17025.

**Comment by feryardiant** [comment]
Just install `ngrok-client` via `apt install` & got same issue

![image](https://user-images.githubusercontent.com/508665/39094217-543a179c-4656-11e8-90e6-123486fc1b55.png)

**Comment by francoisneron** [comment]
+1

**Comment by StoIos** [comment]
+1

**Comment by therealkenc** [comment]
I haven't confirmed the fixininsiderbuilds applied in January, but the +1 static is reason enough to close this. If `sudo npm install -g ngrog` still fails on 17134 or better, reboot under a new issue#.

**Comment by feryardiant** [comment]
It works now.

Thanks team.

**Comment by geetfun** [comment]
On Build 16299.461 and using ngrok version 2.2.8 (direct from ngrok website) on Ubuntu 18.04 seems to fail. 

At this point, is there a recommended version of Windows Build that is confirmed to work?

**Comment by rfgamaral** [comment]
@geetfun Yes, update to the latest 1803 version (April 2018 Update).
